src/nccl_ofi_rdma.c

```diff
@@ -306,65 +263,77 @@ static void unlink_topo_file()
 static int write_topo_file(nccl_ofi_topo_t *topo)
 {
 	int ret = 0;
-	int rc = 0;
-	FILE *file;
-	char *filename;
-	int fd;
+	int topo_fd = -1;
+	int write_fd = -1;
+	FILE *file = NULL;
 
 	/* This function is a no-op in case writing topology file is not enabled explicitly */
 	if (!ofi_nccl_topo_file_write_enable()) {
+		NCCL_OFI_TRACE(NCCL_INIT | NCCL_NET,
+			       "Topology write not enabled; skipping");
 		goto exit;
 	}
 
-	nccl_net_ofi_mutex_lock(&topo_file_lock);
-
-	if (topo_file_unlink) {
-		/* A topology file has already been written and stored
-		 * such that it can be unlinked later. Do not write
-		 * another topology file since it would end up
-		 * overriding the stored filename. */
-		goto unlock;
+	/* If `NCCL_TOPO_FILE` is already set, don't set it again. */
+	if (getenv("NCCL_TOPO_FILE")) {
+		NCCL_OFI_TRACE(NCCL_INIT | NCCL_NET,
+			       "NCCL_TOPO_FILE environment variable is already set to %s",
+			       getenv("NCCL_TOPO_FILE"));
+		goto exit;
 	}
 
-	if (ofi_nccl_topo_file_template()) {
-		filename = strdup(ofi_nccl_topo_file_template());
-	} else {
-		filename = strdup(topo_file_template);
-		/* Store filename to be unlinked later */
-		topo_file_unlink = filename;
+	/* Create file descriptor */
+	topo_fd = memfd_create("ofi_nccl_topo", 0);
+	if (topo_fd == -1) {
+		NCCL_OFI_WARN("Failed to create anonymous topology file. ERROR: %s",
+			      strerror(errno));
+		ret = -errno;
+		goto error;
 	}
 
-	/* Create file descriptor */
-	fd = mkstemp(filename);
-	if (fd == -1) {
-		NCCL_OFI_WARN("Failed to create NCCL topology file from template %s. ERROR: %s",
-			      filename, strerror(errno));
+	/* Duplicate the fd to open the file for writing */
+	write_fd = dup(topo_fd);
```

Comment:
I don't understand why you need to do this?

Reply:
fdopen will not dup the desciptor and fclose closes the underlying descriptor so it kills the descriptor and won't be inherited through fork+exec

Reply:
Yeah, but that's because we're closing the fd.  Duping so we can close the fd is busy work.

Reply:
Yes, duplicating the fd here is inelegant. My thought is that the `FILE *`-style APIs are buffered, so closing the stream will also free the buffer resources, which would not be freed if we just flushed the stream and left it open. The memory footprint is probably trivial, but that's why I preferred closing the stream. I have no strong feeling here.

Another approach I considered was using the unbuffered `read/write` APIs, but the topo writing code is already written to the `FILE *` API (`fprintf`), and I didn't want to change it.

Reply:
`FILE *` interfaces are buffered, which is why there's `fflush()`.

Reply:
Changed to `fflush`